core/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java

```diff
@@ -47,8 +90,42 @@ public DeprecationLogger(ESLogger parentLogger) {
     /**
      * Logs a deprecated message.
      */
-    @SuppressLoggerChecks(reason = "safely delegates to logger")
     public void deprecated(String msg, Object... params) {
-        logger.debug(msg, params);
+        deprecated(THREAD_CONTEXT, msg, params);
     }
+
+    /**
+     * Note: {@code threadContexts} <em>must</em> be a modifiable {@link Set} so that stale contexts can be removed.
```

Comment:
I think it would be simpler to have the node remove its threadcontext from here on close